### ISSUE TITLE
Syntactic ambiguity and leakage in PRONOM signatures.

### ISSUE THREAD

**Issue by nishihatapalmer** [digital-preservation/droid#297]
**Note:** all of what follows is the result of trying to parse PRONOM binary and container signatures, using the rather old spec at https://www.nationalarchives.gov.uk/aboutapps/fileformat/pdf/automatic_format_identification.pdf, and adding in the additional syntax used in container signatures.

**Lexicographic multi byte comparisons**

Various signatures have double byte values in ranges, like the Windows Portable Executable signature:

4D5A{126-128500}50450000{20}0B02{66}**[0000:1000]**

DROID currently interprets this as a set of 00, a range of 00-10, and another 00 byte, which gives a byte range of 0-16. Given this would be more easily represented as just [00:10], I think this is supposed to be matching something else.

The spec says they should be compared lexicographicallly, and also that each should be the same length.  However, DROID doesn't actually do this (and hasn't since at least version 5, not sure what 4 did here).   

So - what should a double-byte range in a signature match?  I guess it should match the spec - but DROID hasn't been doing that for a long time (if it ever did).

**Comment by richardlehane** [comment]
ouch! siegfried will also be bit by this: multi-byte ranges do get parsed as ranges, but the comparison is done lexicographically  (*as the spec says*), using a [go std lib function](https://golang.org/pkg/bytes/#Compare), without interpreting multi-byte strings as ints. If two byte strings are 16 bit ints... do we also have 32 bit and 64 bit ints too? What happens if you have [aaa:bbb] or [a:bb]?

**Comment by nishihatapalmer** [comment]
I've now run my PRONOM parser (for issue https://github.com/digital-preservation/droid/issues/237) on all the signatures for v91.  These are the ones which can't be parsed at present:

fmt/899	4D5A{126-128500}50450000{20}0B01{66}**[0000:1000]**
fmt/900	4D5A{126-128500}50450000{20}0B01{66}**[0000:1000]**
fmt/1105	4244{12}0003{6}**[!&01]**00
fmt/142	52494646{4}57415645666D7420[!10]{3}**[!FEFF]**{16-*}64617461

So we have three things which I can't quite interpret the meaning of, not one:

1.   Range of integer values?   [0000:1000]
2.   Mysterious & symbol         [!&01]               
3.   Not a 16 bit integer           [!FEFF]

**Range of values**
It seems likely that this means a 16 bit range of values, but DROID does not support this.  It will treat this as a range of 00:10 (with the other two 00 bytes added to this set, which means it's still 00:10).

** Mysterious & symbol
The documentation states that this is to only match even values.  This works in DROID, because the **&** operator is a byteseek bitwise operator. that says match all values matching the set bits.  So you get all the odd values with &01.  

I can add support for this to the parser, if TNA have decided to use the & operator as an official part of the PRONOM language.  

Note that the & operator has been deprecated in byteseek 3 - there is alternative more powerful syntax available that can achieve the same thing.  It would be possible to transform any & operator in PRONOM code into the new syntax automatically, so this isn't really a problem.

The real issue here is that TNA need to define the official syntax that PRONOM supports.  Bits of byteseek syntax randomly creeping into PRONOM signatures isn't the ideal way to do this.

**Not a 16 bit integer**
The documentation for this signature states that "wformat Tag not equal to decimal 65534 = 0xFFFE".
There is no support in DROID for matching 16 bit integers (or not 16 bit integers).  This will be interpreted by DROID as not either of the two bytes FF and FE, rather than "not FF followed by FE".

**Comment by nishihatapalmer** [comment]
From testing the PRONOM parser against the existing container signatures, there are also some in there which currently can't be parsed:

 > The character 39 is not a hex digit at position 27 in expression 10000000'Word.Document.'[**'6'-'7'**]00
 > The character 38 is not a hex digit at position 3 in expression [**&01**]
 > [] syntax incorrect. Position: 21 in: 'office:version='[**2227**]'1.0'[**2227**]

So we have:

**String byte ranges**
Need support for specifying byte ranges using ASCII literals, such as '0'-'9'.

**Bitwise & operator**
The **&** operator appears again in a few signatures, matching bits in a byte.

**Multi-byte sets**
The set [2227] is being interpreted (correctly for these signatures) by DROID as a set of two bytes: 0x22 and 0x27, which matches either a double or single quote character.

It is not being interpreted as a 16 bit value 0x2227, and should not be.   This is how byteseek approaches set of bytes (enclosed in square brackets).  The set to match consists of all bytes and byte ranges within the square brackets.  It could also be written (with whitespace) as [22 27].   This is how all regular expression languages use square brackets - all items within the brackets form the set.

So it wouldn't be a bad idea to support multi-byte sets, as that's (a) very useful and (b) completely standard.

**Comment by nishihatapalmer** [comment]
**Range syntax inconsistency**
I've detected another syntactic inconsistency in the signatures.  Binary signatures express ranges of byte values separated by a colon.  Some container signatures use the more standard hyphen for a range of values.  So we have both these in use:
```
30:39
30-39
```
This works because byteseek uses the standard hyphen for ranges, so when it gets passed to the byteseek compiler, it all works (and colons are transformed into hyphens before processing to ensure the standard PRONOM syntax works with byteseek).  This is another example of byteseek syntax leaking into the PRONOM signatures.  

*Options*
In this situation, we  have two options.  First, we could standardise on the colon only (for maximum backwards compatibility), and rewrite those signatures which use the hyphen now.  This means older signature files would not be compatible with later versions of DROID which only use the colon.

Alternatively, we could accept either the colon or hyphen as a valid range syntax (for container signatures at least).   This preserves maximum compatibility with existing and new signatures, at the prices of having two ways of writing the same thing.  

The hyphen is the standard way of expressing ranges in regular expression languages, so it's understandable why people write them this way and my feeling is it should remain supported along with the colon.

**Comment by nishihatapalmer** [comment]
> ouch! siegfried will also be bit by this: multi-byte ranges do get parsed as ranges, but the comparison is done lexicographically (_as the spec says_), using a [go std lib function](https://golang.org/pkg/bytes/#Compare), without interpreting multi-byte strings as ints. If two byte strings are 16 bit ints... do we also have 32 bit and 64 bit ints too? What happens if you have [aaa:bbb] or [a:bb]?

You are right - the spec does say lexicographically.  I've amended the original issue to talk about multi-byte lexicographic comparisons.

**Comment by nishihatapalmer** [comment]
**Signature Leakage**

It's clear that bits of byteseek syntax have leaked into PRONOM syntax, intentionally or otherwise.  It would be better for DROID and PRONOM if the full language, including for container signatures if different, is defined and controlled.

Controls
The validating PRONOM parser written for https://github.com/digital-preservation/droid/issues/237 can help here.  Since it explicitly rejects any syntax or characters it can't understand, it validates that signatures are compliant with the accepted syntactic rules.  

Could build this in somewhere into a validating workflow (possibly inside DROID) that logs warnings or errors for non-compliant signature syntax.

**Comment by DavidUnderdown** [comment]
There's a stored procedure in the PRONOM database that tries to do some consistency checking on various things, including signatures.  It would probably be worth looking at that in the light of these issues too, as it's obviously helpful to stop badly formed signatures getting into the database in the first place, or at least being able to correct them before a new signature file is actually created.  However, that's rathr outside the scope of the current work that's being funded.

**Comment by nishihatapalmer** [comment]
Yes, PRONOM certainly does some consistency checks, but it's not a validating parser.  The one I'm writing does validate the syntax is correct to known rules - so it would be quite easy to make a command line tool (or make DROID warn on improperly formed signatures).

**Comment by nishihatapalmer** [comment]
**Affected Signatures**
Here are all the signatures affected by these issues in one way or another, for both binary and containers.

**Binary Signatures**
fmt/899 		Windows Portable Executable			[0000:1000]	Multi-byte range
fmt/900 		Windows Portable Executable 64 bit		[0000:1000]	Multi-byte range
fmt/1105 		Hierarchical File System				[!&01]		Bitwise &
fmt/142 		Waveform Audio (WAVEFORMATEX)	[!FEFF]		Not byte sequence

**Container Signatures**
fmt/45	Microsoft Word 97 OLE2 Template (OLE2)					[&01]	Bitwise &
fmt/755	Microsoft Word 97 OLE2 Template Password Protected (OLE2)	[&01]	Bitwise &
fmt/754	Microsoft Word 97 OLE2 Password Protected (OLE2)			[&01]	Bitwise &

fmt/136	Open Document Text 1.0 (ZIP)				[22 27]	Double byte set
fmt/290	Open Document Text 1.1 (ZIP)				[22 27]	Double byte set
fmt/291	Open Document Text 1.2 (ZIP)				[22 27] 	Double byte set
fmt/137	Open Document Spreadsheet 1.0 (ZIP)		[22 27] 	Double byte set
fmt/137	Open Document Spreadsheet 1.0 (ZIP)		[22 27] 	Double byte set
fmt/294	Open Document Spreadsheet 1.1 (ZIP)		[22 27] 	Double byte set
fmt/295	Open Document Spreadsheet 1.2 (ZIP)		[22 27] 	Double byte set
fmt/138	Open Document Presentation 1.0 (ZIP)		[22 27] 	Double byte set
fmt/292	Open Document Presentation 1.1 (zip)		[22 27] 	Double byte set
fmt/293	Open Document Presentation 1.2 (ZIP)		[22 27] 	Double byte set
fmt/140	Open Document Database 1.0 (ZIP)			[22 27] 	Double byte set
fmt/444	Open Document Database 1.1 (ZIP)			[22 27] 	Double byte set
fmt/424	Open Document Database 1.2 (ZIP)			[22 27] 	Double byte set

**Comment by Dclipsham** [comment]
Just to flag, we've just published an opportunity (https://www.digitalmarketplace.service.gov.uk/digital-outcomes-and-specialists/opportunities/11061) for a data engineer supplier/specialist to work on next version of PRONOM. Requirements include extending the PRONOM signature vocabulary - https://github.com/nationalarchives/pronom-next/blob/master/PRONOM%20modernisation%20requirements.md

**Comment by nishihatapalmer** [comment]
Great - it will really help to define and extend the PRONOM vocabulary.

In the meantime, I'm going to write the parser so it processes signatures as DROID currently does (even if that is against the old spec), as that works with what we currently have.

We should note that even though DROID doesn't process the multi-byte sequences as per the spec, these signatures still seem to be working, so the real world impact of this seems to be minimal (and it only affects 3 binary signatures).

This means the parser (for the time being) will:

1.  Process multiple bytes and ranges inside square brackets as an arbitrary set of values, not multi-byte sequences.

2.  Include support for the & operator.

3.  Support both the hyphen and colon as a range separator.

I may have a toggle between "strict PRONOM" and "extended Container syntax", but not sure if it's worth it, since the goal of writing the parser is to support PRONOM syntax in container signatures (not to validate PRONOM binary signatures).  This could be added if we want a validating parser for both strict binary and full container syntax.

**Comment by nishihatapalmer** [comment]
**Anchoring Sequence Not To Spec**
There are three signatures which just consist of **[&01]** or [**&01][&01]** 

The spec says that only bytes are allowed to be part of a main anchoring sequence. These bitmask sets should be fragments, and then there is no anchoring sequence left.

DROID has been capable of searching for more complex anchoring sequences for a long time.  It even has code in it to try to grab some of them back from the rather-too-conservative PRONOM signatures, as it can speed up searching by doing this.

**Recommendation**
It doesn't hurt to allow anchors which contain more than just simple bytes - DROID is more than capable of searching for them (and indeed, tries to actively do this any way).

Earlier versions of DROID may not be able to handle these signatures - would require testing.

**Comment by steve-daly** [comment]
@kathaurielle for PRONOM
